From 94ba18672c1638137c9a37e044933d88e410c859 Mon Sep 17 00:00:00 2001
From: Manolo van Ee <manolo.vanee@gmail.com>
Date: Thu, 26 Mar 2026 13:18:10 +0100
Subject: [PATCH 2/4] [stdlib Concurrency] Update Clock.cpp for use with newlib

- Small fixes to make the code more portable.
- Use _POSIX_TIMERS to determine availability of various clock functions.
---
 stdlib/public/Concurrency/Clock.cpp | 25 +++++++++++++------------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/stdlib/public/Concurrency/Clock.cpp b/stdlib/public/Concurrency/Clock.cpp
index 2d1e297b830..ac1c17dc62c 100644
--- a/stdlib/public/Concurrency/Clock.cpp
+++ b/stdlib/public/Concurrency/Clock.cpp
@@ -66,12 +66,12 @@ void swift_get_time(
       continuous.tv_nsec = (interruptTime % 10'000'000) * 100;
 #elif WE_HAVE_STD_CHRONO
       auto now = std::chrono::steady_clock::now();
-      auto epoch = std::chrono::steady_clock::min();
+      auto epoch = std::chrono::steady_clock::time_point::min();
       auto timeSinceEpoch = now - epoch;
       auto sec = std::chrono::duration_cast<std::chrono::seconds>(timeSinceEpoch);
       auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(timeSinceEpoch - sec);
-      continuous.tv_sec = sec;
-      continuous.tv_nsec = ns;
+      continuous.tv_sec = sec.count();
+      continuous.tv_nsec = ns.count();
 #else
 #error Missing platform continuous time definition
 #endif
@@ -101,21 +101,22 @@ void swift_get_time(
       suspending.tv_nsec = (unbiasedTime % 10'000'000) * 100;
 #elif WE_HAVE_STD_CHRONO
       auto now = std::chrono::steady_clock::now();
-      auto epoch = std::chrono::steady_clock::min();
+      auto epoch = std::chrono::steady_clock::time_point::min();
       auto timeSinceEpoch = now - epoch;
       auto sec = std::chrono::duration_cast<std::chrono::seconds>(timeSinceEpoch);
       auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(timeSinceEpoch - sec);
-      suspending.tv_sec = sec;
-      suspending.tv_nsec = ns;
+      suspending.tv_sec = sec.count();
+      suspending.tv_nsec = ns.count();
 #else
 #error Missing platform suspending time definition
 #endif
       *seconds = suspending.tv_sec;
       *nanoseconds = suspending.tv_nsec;
       return;
-    case swift_clock_id_wall:
+    }
+    case swift_clock_id_wall: {
       struct timespec wall;
-#if defined(__linux__) || defined(__APPLE__) || defined(__wasi__) || defined(__OpenBSD__) || defined(__FreeBSD__)
+#if defined(__linux__) || defined(__APPLE__) || defined(__wasi__) || defined(__OpenBSD__) || defined(__FreeBSD__) || defined(_POSIX_TIMERS)
       clock_gettime(CLOCK_REALTIME, &wall);
 #elif defined(_WIN32)
       // This needs to match what swift-corelibs-libdispatch does
@@ -164,7 +165,7 @@ switch (clock_id) {
       auto num = std::chrono::steady_clock::period::num;
       auto den = std::chrono::steady_clock::period::den;
       continuous.tv_sec = num / den;
-      continuous.tv_nsec = (num * 1000000000ll) % den
+      continuous.tv_nsec = (num * 1000000000ll) % den;
 #else
 #error Missing platform continuous time definition
 #endif
@@ -189,7 +190,7 @@ switch (clock_id) {
       auto num = std::chrono::steady_clock::period::num;
       auto den = std::chrono::steady_clock::period::den;
       suspending.tv_sec = num / den;
-      suspending.tv_nsec = (num * 1'000'000'000ll) % den
+      suspending.tv_nsec = (num * 1'000'000'000ll) % den;
 #else
 #error Missing platform suspending time definition
 #endif
@@ -199,7 +200,7 @@ switch (clock_id) {
     }
     case swift_clock_id_wall: {
       struct timespec wall;
-#if defined(__linux__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__wasi__)
+#if defined(__linux__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__wasi__) || defined(_POSIX_TIMERS)
       clock_getres(CLOCK_REALTIME, &wall);
 #elif defined(_WIN32)
       wall.tv_sec = 0;
@@ -240,7 +241,7 @@ void swift_sleep(
     delay = deadline - now;
   }
 #elif defined(__linux__) || defined(__APPLE__) || defined(__wasi__) \
-  || defined(__OpenBSD__) || defined(__FreeBSD__)
+  || defined(__OpenBSD__) || defined(__FreeBSD__) || defined(_POSIX_TIMERS)
   struct timespec ts;
   ts.tv_sec = seconds;
   ts.tv_nsec = nanoseconds;
-- 
2.53.0

